### PR TITLE
KAFKA-13139: Empty response after requesting to restart a connector without the tasks results in NPE

### PR DESCRIPTION
**Pull request by kkonstantine**
Currently tested via the system test `connect_distributed_test.py::ConnectDistributedTest.test_restart_failed_connector`

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by kkonstantine** [comment]
I'd like to check quickly if this can be covered in unit/integration tests as well.

**Comment by kkonstantine** [comment]
cc @kpatelatwork @rhauch @ijuma 
This looks like a blocker for 3.0

**Comment by kpatelatwork** [comment]
@kkonstantine  LGTM .  Thanks a lot for the fixes

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResource.java [hunk 1]
```diff
@@ -269,7 +269,7 @@ public Response restartConnector(final @PathParam("connector") String connector,
             FutureCallback<Void> cb = new FutureCallback<>();
             herder.restartConnector(connector, cb);
             completeOrForwardRequest(cb, forwardingPath, "POST", headers, null, forward);
-            return Response.ok().build();
+            return Response.noContent().build();
```
Comment (rhauch) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResource.java:272]: This returns `204 No Content`, right? But doesn't the old code (prior to this feature) return `200 OK` with no content? 

[KIP-745](https://cwiki.apache.org/confluence/display/KAFKA/KIP-745%3A+Connect+API+to+restart+connector+and+tasks) mentions using `202 Accepted` for the new behavior, but keeps the `200 OK` behavior when `includeTasks=false` and `onlyFailed=false`. 

Are you suggesting that we can't return `200 OK` here if there is no response body and should therefore update the KIP, or can this line be left as is and we change the `RestClient` to properly handle a null content? The issue for this test failure ([KAFKA-111]()) mentions [this code](https://github.com/apache/kafka/blob/d89ea74918ce0a1f2309a09473c9500c52af3b10/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/RestClient.java#L134-L136) in the `RestClient`, which handles all 200 to <300 status codes the same way, so could we just improve that?

**Review by rhauch — COMMENTED** [review]
Thanks for seeing this and proposing the fix. One question about a potentially different fix below.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResource.java (hunk 1 above)
Comment (kkonstantine) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResource.java:272]: Thanks for the review @rhauch. I had the same questions and concerns. 

Actually, the current behavior for both connector and task restarts is to return `HTTP/1.1 204 No Content`. So this fix should retain this behavior for the connector restart. 

This seemed the most appropriate thing to do for a POST request whose response would have an empty body given the description in the RFC:
https://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html#sec10.2.5

I also think that an addendum to the KIP would be a good idea one we finish the review here. 
Wdyt?

Comment (kpatelatwork) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResource.java:272]: @rhauch and @kkonstantine  I confirm that from AK2.8 test the old behavior was returning 204, We would need to fix the documentation and KIP. Please see below

> ➜  kafka_2.13-2.8.0 curl -v -X POST http://localhost:8083/connectors/local-file-source/restart
> *   Trying ::1...
> * TCP_NODELAY set
> * Connected to localhost (::1) port 8083 (#0)
> > POST /connectors/local-file-source/restart HTTP/1.1
> > Host: localhost:8083
> > User-Agent: curl/7.64.1
> > Accept: */*
> > 
> < HTTP/1.1 204 No Content
> < Date: Mon, 26 Jul 2021 23:41:44 GMT
> < Server: Jetty(9.4.39.v20210325)
> < 
> * Connection #0 to host localhost left intact
> * Closing connection 0
> ➜  kafka_2.13-2.8.0

Comment (rhauch) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResource.java:272]: Awesome. Thanks, folks!

**Review by rhauch — APPROVED** [review]
LGTM, especially since we've verified this fix matches and restores the 2.x behavior.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResource.java (hunk 1 above)
Comment (rhauch) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResource.java:272]: I've corrected the KIP and sent an email describing this minor correction to the vote thread for the KIP.

I've also added [a comment on KAFKA-13139](https://issues.apache.org/jira/browse/KAFKA-13139?focusedCommentId=17388116&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-17388116) that describes the root cause and the KIP correction.

Thanks, @kkonstantine and @kpatelatwork.

**Comment by kkonstantine** [comment]
Thank you both. Merging to trunk and 3.0

**Comment by kkonstantine** [comment]
A single, non-relevant, test failed on the second build: 
`Build / JDK 8 and Scala 2.12 / kafka.server.DelegationTokenRequestsTest.testDelegationTokenRequests()`

**Pull request merged**